atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java

```diff
@@ -0,0 +1,149 @@
+import com.google.common.io.ByteStreams;
+
+import net.jpountz.lz4.LZ4BlockOutputStream;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.xxhash.XXHashFactory;
+
+/**
+ * {@link InputStream} that wraps a delegate InputStream, compressing its
+ * contents as they are read. Compression is managed via a compressing
+ * {@link OutputStream}. Reads pull data as necessary from the delegate
+ * InputStream, write the uncompressed data through the OutputStream, then
+ * serve the read from the resulting compressed buffer.
+ */
+public final class LZ4CompressingInputStream extends BufferedDelegateInputStream {
+
+    private static final LZ4Compressor COMPRESSOR = LZ4Factory.fastestInstance().fastCompressor();
+    private static final int DEFAULT_SEED = 0x9747b28c;
+    private static final int DEFAULT_BLOCK_SIZE = 1 << 16; // 64 KB
+    private static final int LZ4_HEADER_SIZE = 21;
+
+    private final LZ4BlockOutputStream compressingStream;
+    private final int blockSize;
+    private final byte[] uncompressedBuffer;
+
+    // Position in the compressed buffer while writing
+    private int writeBufferPosition;
+    // Flag to indicate whether this stream has been exhausted.
+    private boolean finished;
+
+    public LZ4CompressingInputStream(InputStream delegate) throws IOException {
+        this(delegate, DEFAULT_BLOCK_SIZE);
+    }
+
+    public LZ4CompressingInputStream(InputStream delegate, int blockSize) throws IOException {
+        super(delegate, LZ4_HEADER_SIZE + COMPRESSOR.maxCompressedLength(blockSize));
+        this.blockSize = blockSize;
+        this.uncompressedBuffer = new byte[blockSize];
+        Checksum checksum = XXHashFactory.fastestInstance().newStreamingHash32(DEFAULT_SEED).asChecksum();
+        OutputStream delegateOutputStream = new InternalByteArrayOutputStream();
+        this.compressingStream = new LZ4BlockOutputStream(delegateOutputStream, blockSize, COMPRESSOR, checksum, true);
+        this.finished = false;
+    }
+
+    @Override
+    protected int refill() throws IOException {
+        if (finished) {
+            return 0;
+        }
+        writeBufferPosition = 0;
+
+        // Read a block of data from the delegate input stream.
+        int bytesRead = ByteStreams.read(delegate, uncompressedBuffer, BUFFER_START, blockSize);
+        if (bytesRead == 0) {
+            // The delegate input stream has been exhausted, so we notify
+            // the compressing stream that there are no remaining bytes.
+            // Lastly, we flush the LZ4 footer into the internal buffer.
+            compressingStream.finish();
+            compressingStream.flush();
+            finished = true;
+        } else {
+            // Write the bytes to the compressing stream and flush it.
+            compressingStream.write(uncompressedBuffer, BUFFER_START, bytesRead);
+            compressingStream.flush();
+        }
+
+        return writeBufferPosition;
```

Comment:
Calling compressingStream.flush writes data through to the InternalByteArrayOutputStream and the write methods there update writeBufferPosition.